### PR TITLE
kaggle extraction integration:  successful (50 models tested)

### PR DESCRIPTION
**Pull request by Dhwanisolanki12**
# Add Kaggle extractor (extraction + enrichment, 1 commit), Integrating model instances entity identifier, 2nd commit)

Adds Kaggle as a platform in the ETL pipeline, following the AI4Life structure:
a thin extractor delegating to per-entity clients, entity identifiers for
related entities, and a Dagster asset graph rooted at a timestamped run folder.

## Extraction

**`kaggle_run_folder`** → **`kaggle_model_refs`** → **`kaggle_raw_records`** → **`kaggle_models_raw`**
plus **`kaggle_raw_catalog_sources`** for provenance.

Kaggle differs from the other platforms in one important way: there is no
endpoint that returns many models at once. Every model card needs its own API
call, and there are ~43,000 of them. `kaggle_model_refs` derives the model list
by joining Meta Kaggle's `Models.csv` against `Organizations.csv` and
`Users.csv`; `kaggle_raw_records` then crawls the API for each ref.

That crawl is handled by `KaggleCrawler`, which is the one file with no AI4Life
analogue. It provides parallel fetching, adaptive rate limiting, crash-resume,
and incremental refresh. Keeping it separate leaves `KaggleExtractor` the same
shape and size as `AI4LifeExtractor`.

### Performance

Full extraction runs in **~6 hours**, down from ~2 days for a naive sequential
crawl — roughly a 10× improvement. The remaining limit is Kaggle's, not ours:
the API enforces a rolling hourly quota of roughly 7–9k records on this
endpoint. This was measured rather than assumed — throttling occurs even at 0.5
req/s, which rules out a per-second speed cap, and several configurations all
converged on the same hourly total. Rate limits are therefore class constants in
`KaggleCrawler` rather than config values, with a comment explaining why raising
them does not help.

Data quality on the full run: 22 failures out of ~43,000 (0.05%), all deleted or
renamed models. 600+ rate-limit events, all automatically retried and recovered,
zero records lost.

### Resume and incremental

Records are appended to an NDJSON log and fsynced every 200 rows, so an
interrupted run continues from its last checkpoint rather than starting over.
This was validated in production during a mid-crawl restart at 10,500 records
with no data loss.

Subsequent runs fingerprint Meta Kaggle's date/version columns and fetch only
new or changed models — typically ~1,000 records, about 10 minutes instead of 6
hours.

Because resume and incremental both depend on reading what previous runs
collected, that state lives under `CACHE_PATH` (`/data/cache/kaggle/`) rather
than inside the run folder. A per-run location would reset it every time and
turn every run into a fresh multi-hour crawl. Run folders under
`/data/1_raw/kaggle/<timestamp>_<uuid>/` still receive the full output of each
run: `records.json`, `models.json`, `instances.json`, `keywords.json`,
`sources.json`.

## Enrichment

Two entity types, both following the `identified_*` → `*_raw` pattern used for
AI4Life datasets and keywords.

### Instances

A Kaggle model is a container; the downloadable artifacts are its *instances*,
one per framework and variation. `google/bert` has 31, each with its own
license, version, size and URL. Flattening those onto the model row would lose
which framework belongs to which variation, so they are emitted as their own
entities linked back by `parent_mlentory_id`.

Instance ids mirror the Kaggle URL path
(`google/bert/TensorFlow2/answer-equivalence-bem`) so they round-trip to a real
page. This matters: Kaggle's `framework` field is inconsistent across records —
`pyTorch`, `keras`, and `MODEL_FRAMEWORK_TENSOR_FLOW_2` all appear — while the
URL always carries the display form. Both are kept, `modelArchitecture` for the
raw value and `frameworkName` for the display form.

Every raw instance field is passed through alongside the schema-aligned ones, so
fields Kaggle adds later are not silently dropped. This already surfaced
`baseModelInstanceInformation`, a nested object carrying model lineage for
fine-tuned instances, which is worth mapping into the graph in a follow-up.

#### Kaggle Configuration
 ```
 kaggle:
    # Number of models to extract. null = all (~43,044)
    num_models: null
    # Fetch only new/changed models since the last run.
    # Keep false for the baseline crawl; switch to true afterwards.
    incremental: false
    # Force a full refetch regardless of state
    force_full_refresh: false
    # Re-download Meta Kaggle CSVs to pick up new models (daily refresh)
    refresh_metadata: true
    # Parallel threads. Throughput is capped by Kaggle's hourly quota,
    # not by this value - raising it does not make the crawl faster.
    threads: 8
```

#### Output files are below to review:
[instances.json](https://github.com/user-attachments/files/31034535/instances.json)
[Kaggle_models.json](https://github.com/user-attachments/files/31034537/Kaggle_models.json)
[records.json](https://github.com/user-attachments/files/31034538/records.json)

### Models and instances in one file

`mlmodels.json` now holds both models and their instances as separate
`MLModel` records, rather than splitting instances into `mlinstances.json`.
A Kaggle model is a container and its instances are the downloadable
artifacts — one per framework and variation — and each is a model in its own
right.

Instances are identifiable by a populated `baseModel`: on an instance it
points at the model it belongs to, and additionally at the model it was
derived from where Kaggle records real lineage. Parent models have an empty
`baseModel`, since Kaggle records lineage per instance only.

The `kaggle_instances_normalized` and `kaggle_load_instances_to_neo4j` assets
are removed; instance normalization now happens inside
`kaggle_model_normalized`.

Note this means models and instances share the `fair4ml:MLModel` type in the
graph. `schema:isPartOf` would separate "packaged from" from "derived from"
properly, but that needs a line in the shared `rdf_loader.build_model_triples`
property list — raised separately rather than done here.

### `adaptionTechniques`

Added to both models and instances, set to `"fineTuned"` where Kaggle's
`fineTunable` flag is present and `None` otherwise. The flag is declared per
instance, so parent models resolve to `None`.

### `sharedBy` entity

New identifier and client pair, following the same shape as licenses,
keywords, and frameworks. On Kaggle this is the account that uploaded the
model — a user or an organization — so grouping by it answers "show me
everything this account published".

Entities are de-duplicated by name: one node, many edges. The value stored on
a model is the minted `mlentory_id`, not the display name; the readable form
comes from `translation_mapping.json`, matching how the HF pipeline renders
it.

Instances inherit their parent's `sharedBy` rather than being left empty, so a
search for an account's work returns the variations alongside the models. The
IRI is minted in the merge step for instances, since entity linking only
covers `models.json` — verified byte-identical to the model's, so both point
at one node rather than a name and an IRI diverging.

Two details worth noting from testing: pandas turns missing cells into float
`nan`, whose `str()` is `"nan"` and would otherwise be minted as a real
entity, so it is filtered explicitly. And `parent_name` is deliberately not a
fallback source — on an instance row it holds the parent model's *title*, not
its owner, so using it would create entities like "Phi-3" as if they were
people.

### `metrics` removed

Dropped from both models and instances. It held platform counters (votes,
downloads, instance count, version) that no downstream consumer reads, and
`build_model_triples` does not map the field, so nothing in it reached the
graph.

### Instance display names

An instance whose slug is generic — `default`, `1`, or empty, which covers
every single-variation model — now takes its name from the parent title plus
framework: `Simple CNN (PyTorch)` rather than one of forty identical
`default` entries. Real variation names like `phi-3.5-mini-instruct` pass
through untouched, and the original slug is preserved in the `slug` field.

### Framework name normalization

Kaggle reports the framework inconsistently across records — `pyTorch`,
`keras`, and `MODEL_FRAMEWORK_TENSOR_FLOW_2` all appear — while the instance
URL always carries the display form. `models_client` now writes a `frameworks`
array read from those URLs, so `PyTorch`, `Keras`, and `TensorFlow2` each
resolve to a single entity instead of several spellings of the same one.

The same normalization applies when reconstructing base-model references,
where `baseModelInstanceInformation` uses the field spelling: without it the
minted IRI would not match the instance it points at.

### Multi-valued license handling

License belongs to the instance, and one model can carry instances under
different licenses. `models_client` was joining them into a single display
string, so `"MIT, Apache 2.0"` would have been identified as one license.
It now also writes a `licenses` array, which the identifier reads. Splitting
the display string on a comma is deliberately avoided — license names
legitimately contain punctuation, e.g.
`Attribution-NonCommercial-ShareAlike 3.0 IGO (CC BY-NC-SA 3.0 IGO)`.

[mlmodels.json](https://github.com/user-attachments/files/31298879/mlmodels.json)

### CODE REVIEW

**Comment by ShriSuha** [comment]
**The Kaggle pipeline is well structured and works perfectly fine with no errors.**

I made a few necessary follow-up changes:

1) Load sharedBy into Neo4j - extract already minted the IRIs; they were not loaded
2) Pipeline tag Kaggle_etl → kaggle_etl, plus Makefile targets (kaggle-etl, kaggle-extract, kaggle-transform, kaggle-load, kaggle-index, kaggle-vector)
3) inLanguage property added - same Lingua / pycountry path as AI4Life
4) Instance licenses - mint W3ID IRIs instead of plain names
5) Vector backfill asset added - kaggle_vector_backfill was missing for the vector search
